### PR TITLE
ONNX export: Add Crop, Deconvolution and fix the default stride of Pooling to 1

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
Add operators to ONNX exporter:
 - Crop
 - Deconvolution
Fix pooling default stride to 1.
Fixes: https://github.com/apache/incubator-mxnet/issues/12807
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by vandanavk** [comment]
@ptrendx Did you get a chance to test these operators with ONNX? onnx_backend_test.py

**Comment by Roshrini** [comment]
@ptrendx Thanks for submitting this PR. :)
You should add 'test_ConvTranspose2d' in "Basic Model Tests" to make sure "Deconvolution" operator is running fine in this file. https://github.com/apache/incubator-mxnet/blob/master/tests/python-pytest/onnx/export/onnx_backend_test.py

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 1]
```diff
@@ -207,6 +207,89 @@ def convert_convolution(node, **kwargs):
...
+        bias_node = proc_nodes[kwargs["index_lookup"][inputs[2][0]]].name
+
+    attrs = node.get("attrs")
+
+    kernel_dims = list(parse_helper(attrs, "kernel"))
+    stride_dims = list(parse_helper(attrs, "stride", [1, 1]))
+    pad_dims = list(parse_helper(attrs, "pad", [0, 0]))
+    num_group = int(attrs.get("num_group", 1))
+    dilations = list(parse_helper(attrs, "dilate", [1, 1]))
+    adj_dims = list(parse_helper(attrs, "adj"))
```
Comment (Roshrini) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Can you please add default value here?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 2]
```diff
@@ -207,6 +207,89 @@ def convert_convolution(node, **kwargs):
...
+        pads=pad_dims,
+        group=num_group,
+        name=name
+    )
+
+    return [deconv_node]
+
+
+@mx_op.register("Crop")
+def convert_crop(node, **kwargs):
```
Comment (Roshrini) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Crop is an experimental operator in ONNX, its definition can be changed. Suggest: add warning for users to be cautious while using this op?

**Comment by sandeep-krishnamurthy** [comment]
@ptrendx - ping :-)

**Comment by vandanavk** [comment]
ONNX 1.3 has been merged. Please rebase this PR. Thanks.

**Comment by Roshrini** [comment]
@ptrendx Can you please address review comments and rebase this PR? Thanks.

**Comment by ptrendx** [comment]
Hi, I'm sorry for lack of communication on my part. Unfortunately I got pulled into other high priority work and will be able to revisit this PR afterwards.

**Comment by roywei** [comment]
Hi @ptrendx , were you able to get some time on this PR? Do you need any help to continue with this?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 2 above)
Comment (ptrendx) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: done.

**Comment by ptrendx** [comment]
Hi, I'm sorry for the long delay in answering the comments. It's rebased onto the latest master.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 3]
```diff
@@ -219,6 +219,85 @@ def convert_convolution(node, **kwargs):
...
+    """Map MXNet's deconvolution operator attributes to onnx's ConvTranspose operator
+    and return the created node.
+    """
+    name, inputs, attrs = get_inputs(node, kwargs)
+
+    num_inputs = len(inputs)
+
+    proc_nodes = kwargs["proc_nodes"]
+    input_node = proc_nodes[kwargs["index_lookup"][inputs[0][0]]].name
+    weights_node = proc_nodes[kwargs["index_lookup"][inputs[1][0]]].name
```
Comment (vandanavk) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: input_nodes returned by get_inputs() already has all the node names. Lines 231-236, 247-249 are not required on the latest code. directly pass input_nodes in make_node

Comment (ptrendx) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Done.

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 4]
```diff
@@ -219,6 +219,72 @@ def convert_convolution(node, **kwargs):
...
+def convert_crop(node, **kwargs):
+    """Map MXNet's crop operator attributes to onnx's Crop operator
+    and return the created node.
+    """
+    name, inputs, attrs = get_inputs(node, kwargs)
+
+    num_inputs = len(inputs)
+
+    proc_nodes = kwargs["proc_nodes"]
+    input_node = proc_nodes[kwargs["index_lookup"][inputs[0][0]]].name
```
Comment (vandanavk) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: input_node is returned by get_inputs

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 5]
```diff
@@ -571,7 +637,7 @@ def convert_pooling(node, **kwargs):
 
     kernel = eval(attrs["kernel"])
     pool_type = attrs["pool_type"]
```
Comment (Roshrini) [inline python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Can you also add default value for pool_type?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py [hunk 6]
```diff
@@ -219,6 +219,72 @@ def convert_convolution(node, **kwargs):
...
+        scale=[1, 1],
+        name=name
+    )
+
+    logging.warning(
+        "Using an experimental ONNX operator: Crop. " \
+        "Its definition can change.")
+
+    return [crop_node]
+
```
[the remark this answers is not in the document]
Comment (Roshrini) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: +1

**Comment by Roshrini** [comment]
@ptrendx Can you address the review comments so that we can merge this PR? :)

**Comment by Roshrini** [comment]
@ptrendx Can you please rebase this PR?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 4 above)
Comment (ptrendx) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Done

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ptrendx) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: @vandanavk Could you advise where should I put the test for just export? I made the test but am unsure where should it go after the refactor. My understanding is that test_nodes.py is for things that can be both exported and imported, should I put it in mxnet_export_test.py then?

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 5 above)
Comment (ptrendx) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: Done

File: python/mxnet/contrib/onnx/mx2onnx/_op_translations.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vandanavk) [reply python/mxnet/contrib/onnx/mx2onnx/_op_translations.py]: @ptrendx export tests usually perform an import-export-import on an ONNX model and then compare inference results.

if you are planning to add an ONNX import for crop, then adding the test would be straightforward - just add a test_case in test_node.py.

if you are adding export alone for crop, then in test_node.py, create a new test case list (same format as the existing maybe), add a new function test_export() in class TestNode(unittest.TestCase), export to onnx format. not sure how to test inference in this case though.

**Comment by stu1130** [comment]
@Roshrini could you take a look at this?Thanks

**Comment by Roshrini** [comment]
Hi @ptrendx, if you have test ready for crop, can you please add it? Or maybe create a separate PR for crop operator? Other part of this PR looks good and is ready to be merged.

**Comment by Roshrini** [comment]
Thank you for working on this ops @ptrendx :) I will add test case for crop in my PR as I am already adding tests for export ops in this PR: https://github.com/apache/incubator-mxnet/pull/13981

**Pull request merged**